README.md

```diff
@@ -1,229 +1,894 @@
+present it as a `__sys_bpf` -> `map_create` -> `find_and_alloc_map` call chain,
+which is what is logically happening according to C source code.
+
+To the right of most functions we can see source code line location that was
+executing at the time of capturing stack trace. `kernel/bpf/syscall.c:1106:8`
+(next to `. map_create`) points out that C source code line #1106 in
+`kernel/bpf/syscall.c` file was the one being executed at the time of
+`find_and_alloc_map` -> `array_map_alloc` -> `bpf_map_alloc_percpu` call chain
+executing inside `map_create`. In source code corresponding to the kernel I've
+run this maps correctly to:
+
+```c
+  1105        /* find map type and init map: hashtable vs rbtree vs bloom vs ... */
+  1106        map = find_and_alloc_map(attr);
+  1107        if (IS_ERR(map))
+  1108                return PTR_ERR(map);
+```
+
+To the left of each function that was specified through `-e` and `-a` arguments
+you'll also see two more pieces of information: function latency (e.g., 73
+microseconds for `__x64_sys_bpf`) and it's return result, which could be one of
+`-Exxx` error codes, `NULL`, `true`/`false`, some valid address like
+`0xffffc90000243000`, some other integer value, or even `void`; `retsnoop` will
+do its best to represent captured return value as meaningfully as possible,
+based on captured return value and function's type signature, if available. In
+our case, we see that innermost call to `bpf_map_alloc_percpu` returned `NULL`,
+which the got translated to `-ENOMEM` error and bubbled up all the way through
+`__x64_sys_bpf` and eventually got returned to user-space as a result of
+`bpf()` syscall.
+
+There could be few extra markers like '!', '*', etc, at the very left side of a
+line, but we won't go into detail on what they mean. Usually it's pretty
+harmless to ignore them.
+
+So, in general, what did we achieve and what can we understand from
+`retsnoop`'s output? With a single invocation of `retsnoop` with few arguments
+we managed to trace `bpf()` syscall all the way down to `bpf_map_alloc_percpu`
+function returning NULL. If we dig deeper into source code, we'll see that it
+further calls `__alloc_percpu_gfp` (which we didn't instruct `retsnoop` to
+trace, as it wasn't in either `-e` nor `-a` arguments, so we can't see that
+from the above stack trace) which got to be the one further returning `NULL`,
+indicating allocation failure. We can keep digging deeper to try to understand
+exactly why, by reading code and adjusting `retsnoop`'s set of traced function
+, but that's beyond the scope of this example.
+
+The main point is that we can see the sequence of function calls
+leading all the way to an attempt to allocate memory for per-CPU BPF array map,
+and that it is exactly `bpf_map_alloc_percpu` that's failing. We also know that
+`NULL` got translated into `-ENOMEM` before bubbling up to user-space as an
+error returned from `bpf()` syscall.  Without `retsnoop` it would be a
```

Comment:
s/from `bpf()`/from the `bpf()`

also, I suggest removing 'a' from the end of the line, and rephrasing the sentence to:


...it would be considerably harder to narrow it down so precisely...